### PR TITLE
Twisted SSL context and cloud support

### PR DESCRIPTION
**Pull request by TheRealFalcon**
Might want to review the two commits separately.

Ran all tests (including long and cloud) using Twisted on a separate branch:
https://jenkins-drivers.build.dsinternal.org/job/riptano.python-driver.twisted-cloud.twisted_cloud/6/

The Unix socket test fails on twisted on master and this PR doesn't address that.

My reasoning behind the SSL context changes...

For the SSL verification, we need to:
1. Verify we have a valid certificate
2. Verify the hostname is correct

When we call `set_verify(VERIFY_PEER, callback)`, number 1 happens, but we still control what happens with the result:

> The return value of verify_callback controls the strategy of the further verification process. If verify_callback returns 0, the verification process is immediately stopped with "verification failed" state. If SSL_VERIFY_PEER is set, a verification failure alert is sent to the peer and the TLS/SSL handshake is terminated. If verify_callback returns 1, the verification process is continued. If verify_callback always returns 1, the TLS/SSL handshake will not be terminated with respect to verification failures and the connection will be established.

^ is from https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_verify.html

The documentation does also state that the callback can be a null pointer and a default one will be used, but I'm not sure how that's been translated into python and I haven't seen any python examples that just pass `None`, so I'm sticking with an us-defined callback for now. So regardless of any sort of hostname verification (number 2 that I listed above), we still need the `set_verify` call and the verification callback for number 1 to happen.

In the past, we also did do number 2 (ourselves manually) in this verification callback, and that's the most logical place to do it. But the fact that this verification callback was also tied to the verify method (e.g., SSL_VERIFY_PEER vs SSL_VERIFY_NONE in the same documentation above) made allowing a user-created SSL context problematic. We want to allow them to set the verify method, but disallow them from doing the hostname verification so we can do it ourselves. Doing our verification in `set_info_callback` was a convenient way to make that happen.
https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_info_callback.html
This is where Twisted normally does its hostname verification too:
https://github.com/twisted/twisted/blob/twisted-19.7.0/src/twisted/internet/_sslverify.py#L1205

So to summarize, we still need `set_verify` and the verification callback in order to ensure we have a valid certificate, but now we can do hostname verification in the `set_info_callback` so that we don't interfere with a user-provided ssl context.

### CODE REVIEW

File: cassandra/io/twistedreactor.py [hunk 1]
```diff
@@ -166,47 +133,46 @@ def _on_loop_timer(self):
...
+                )
+        self.context.set_info_callback(self.info_callback)
 
-    _HAS_SSL = True
+    def verify_callback(self, connection, x509, errnum, errdepth, ok):
+        return ok
 
-except ImportError as e:
-    _HAS_SSL = False
+    def info_callback(self, connection, where, ret):
```
Comment (TheRealFalcon) [inline cassandra/io/twistedreactor.py]: Borrowed the idea from here:
https://github.com/twisted/twisted/blob/twisted-19.7.0/src/twisted/internet/_sslverify.py#L1225

File: cassandra/io/twistedreactor.py [hunk 2]
```diff
@@ -246,29 +212,48 @@ def __init__(self, *args, **kwargs):
...
+        Convenience function to connect and store the resulting
+        connector.
+        """
+        host, port = self.endpoint.resolve()
+        if self.ssl_context or self.ssl_options:
+            # Can't use optionsForClientTLS here because it *forces* hostname verification.
+            # Cool they enforce strong security, but we have to be able to turn it off
+            self._check_pyopenssl()
+
+            ssl_options = SSLCreator(
```
Comment (aboudreault) [inline cassandra/io/twistedreactor.py]: nit: should this be named `ssl_connection_creator`

File: cassandra/io/twistedreactor.py [hunk 3]
```diff
@@ -246,29 +212,48 @@ def __init__(self, *args, **kwargs):
...
+                self.endpoint.address,
+                self.ssl_context if self.ssl_context else None,
+                self.ssl_options,
+                self._check_hostname,
+                self.connect_timeout,
+            )
+
+            point = SSL4ClientEndpoint(
```
Comment (aboudreault) [inline cassandra/io/twistedreactor.py]: nit: `endpoint`

File: cassandra/cluster.py [hunk 4]
```diff
@@ -36,6 +36,9 @@
 
 import weakref
 from weakref import WeakValueDictionary
+
+from cassandra.io.twistedreactor import TwistedConnection
+
```
Comment (aboudreault) [inline cassandra/cluster.py]: crash if twisted is not installed.

File: cassandra/datastax/cloud/__init__.py [hunk 5]
```diff
@@ -165,3 +166,17 @@ def _ssl_context_from_cert(ca_cert_location, cert_location, key_location):
...
     return ssl_context
+
+
+def _pyopenssl_context_from_cert(ca_cert_location, cert_location, key_location):
+    try:
+        from OpenSSL import SSL
+    except ImportError:
+        return None
+    ssl_context = SSL.Context(SSL.TLSv1_METHOD)
+    ssl_context.set_verify(SSL.VERIFY_PEER, callback=lambda _1, _2, _3, _4, ok: ok)
```
Comment (aboudreault) [inline cassandra/datastax/cloud/__init__.py]: sorry, still trying to understand the set_verify use. Maybe a good thing to summarize set_verify vs info_callback in the PR summary.

File: cassandra/datastax/cloud/__init__.py [hunk 6]
```diff
@@ -117,6 +117,7 @@ def parse_cloud_config(path, cloud_config):
         cert_location = os.path.join(config_dir, 'cert')
         key_location = os.path.join(config_dir, 'key')
         config.ssl_context = _ssl_context_from_cert(ca_cert_location, cert_location, key_location)
+        config.pyopenssl_context = _pyopenssl_context_from_cert(ca_cert_location, cert_location, key_location)
```
Comment (aboudreault) [inline cassandra/datastax/cloud/__init__.py]: wondering if pyopenssl is a dependency of a lot of python packages (server, web etc.)... would be a nice to have to just create the ssl context we need.

File: cassandra/io/twistedreactor.py [hunk 7]
```diff
@@ -166,47 +133,46 @@ def _on_loop_timer(self):
...
+    def verify_callback(self, connection, x509, errnum, errdepth, ok):
+        return ok
 
-except ImportError as e:
-    _HAS_SSL = False
+    def info_callback(self, connection, where, ret):
+        if where & SSL.SSL_CB_HANDSHAKE_DONE:
+            if self.check_hostname and self.host != connection.get_peer_certificate().get_subject().commonName:
+                transport = connection.get_app_data()
+                transport.failVerification(Failure(Exception("Hostname verification failed")))
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cassandra/io/twistedreactor.py]: Sure. I wasn't sure the exact exception to use here.

File: cassandra/io/twistedreactor.py [hunk 8]
```diff
@@ -246,29 +212,48 @@ def __init__(self, *args, **kwargs):
...
-        connector.
-        """
+    def _check_pyopenssl(self):
         if self.ssl_options:
-
             if not _HAS_SSL:
                 raise ImportError(
-                    str(e) +
+                    str(import_exception) +
                     ', pyOpenSSL must be installed to enable SSL support with the Twisted event loop'
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cassandra/io/twistedreactor.py]: No, it's actually not unfortunately. Users will need to install it themselves if they want to use twisted and ssl

File: cassandra/io/twistedreactor.py (hunk 2 above)
Comment (TheRealFalcon) [reply cassandra/io/twistedreactor.py]: Sure. It's a replacement for the `clientForTlsOptions` call which is why I named it that, but the parameter it takes in the `SSL4ClientEndpoint` call is called `sslContextFactory`. Twisted is weird. I'll just name it `ssl_connection_creator` :D

File: cassandra/io/twistedreactor.py [hunk 9]
```diff
@@ -246,29 +212,48 @@ def __init__(self, *args, **kwargs):
...
+        connector.
+        """
+        host, port = self.endpoint.resolve()
+        if self.ssl_context or self.ssl_options:
+            # Can't use optionsForClientTLS here because it *forces* hostname verification.
+            # Cool they enforce strong security, but we have to be able to turn it off
+            self._check_pyopenssl()
+
+            ssl_options = SSLCreator(
+                self.endpoint.address,
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cassandra/io/twistedreactor.py]: Yes. The address here will contain the hostname rather than the resolved address which is what we'll need to check on in the ssl handshake.

File: cassandra/datastax/cloud/__init__.py (hunk 6 above)
Comment (TheRealFalcon) [reply cassandra/datastax/cloud/__init__.py]: So are you saying rather than create both the standard ssl context and the pyopenssl context, you'd rather make a decision in the cloud config parsing code and just set a single context there? I considered that but thought this might be a little cleaner but also have no problem doing it that way. Would you rather I just set a single context?

File: cassandra/io/twistedreactor.py [hunk 10]
```diff
@@ -166,47 +133,46 @@ def _on_loop_timer(self):
...
-
-        def verify_callback(self, connection, x509, errnum, errdepth, ok):
-            if ok:
-                if self.check_hostname and self.host != x509.get_subject().commonName:
-                    return False
-            return ok
+                self.context.set_verify(
+                    self.ssl_options["cert_reqs"],
+                    callback=self.verify_callback
+                )
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cassandra/io/twistedreactor.py]: 1. `verify_callback` is still returning the result of the pre-verification. I'll put more info in the PR description.
2. `cert_reqs` is still there. I just changed indentation :)

**Comment by TheRealFalcon** [comment]
@aboudreault I updated the PR description with more info about the SSL context and the callbacks being used.

File: cassandra/io/twistedreactor.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (aboudreault) [reply cassandra/io/twistedreactor.py]: ah, forgot that twisted was handling the cert_reqs internally.

File: cassandra/datastax/cloud/__init__.py (hunk 6 above)
Comment (aboudreault) [reply cassandra/datastax/cloud/__init__.py]: I don't care much if we have clearer and different ssl context attributes... I care more about constructing both when we can determine easily which one we need.

File: cassandra/io/twistedreactor.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (aboudreault) [reply cassandra/io/twistedreactor.py]: I understand what is verify_callback yes, just not sure about the point to create a verify_callback that just return OK.

File: cassandra/io/twistedreactor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (aboudreault) [reply cassandra/io/twistedreactor.py]: ah.. right... that's why that we had twisted[tls] in test-requirements..

File: cassandra/io/twistedreactor.py (hunk 2 above)
Comment (aboudreault) [reply cassandra/io/twistedreactor.py]: damn twisted and their 234 compatible interfaces

**Comment by aboudreault** [comment]
Ok... I understand the why of the empty verify_callback now. Thanks for the explanation. It is indeed not possible to pass a null pointer/None with Python: https://github.com/pyca/pyopenssl/blob/master/src/OpenSSL/SSL.py#L1104 :+1:

File: cassandra/io/twistedreactor.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cassandra/io/twistedreactor.py]: OK is a variable. It'll be true if everything checks out and false if not.

[the remark this answers is not in the document]
Comment (aboudreault) [reply cassandra/io/twistedreactor.py]: Yes I know ;)

[the remark this answers is not in the document]
Comment (aboudreault) [reply cassandra/io/twistedreactor.py]: See my last comment, we are good. Sorry for the confusion if I was clear. I know the verify_callback has a use... I was just thinking it is bad that we need to set a callback that does nothing. By *nothing* I mean *doesn't do any verification*.

File: cassandra/io/twistedreactor.py [hunk 11]
```diff
@@ -55,62 +65,20 @@ def dataReceived(self, data):
...
-        in the ClientFactory.
-        """
-        return TwistedConnectionProtocol()
-
-    def clientConnectionFailed(self, connector, reason):
-        """
-        Overridden twisted callback which is called when the
-        connection attempt fails.
-        """
-        log.debug("Connect failed: %s", reason)
```
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cassandra/io/twistedreactor.py]: I don't think there's a suitable analog. The protocol interface doesn't have a `connectionFailed` method, just `connectionLost`.

[the remark this answers is not in the document]
Comment (aboudreault) [reply cassandra/io/twistedreactor.py]: Okay

File: test-requirements.txt [hunk 12]
```diff
@@ -15,3 +15,4 @@ cython>=0.20,<0.30
 packaging
 futurist; python_version >= '3.7'
 asynctest; python_version > '3.4'
+pyopenssl
```
[the remark this answers is not in the document]
Comment (aboudreault) [reply test-requirements.txt]: ^

[the remark this answers is not in the document]
Comment (aboudreault) [reply test-requirements.txt]: ah, this command has the OUTDATED flag :unamused:  resolving

File: docs/security.rst [hunk 13]
```diff
@@ -239,6 +248,28 @@ The following driver code specifies that the connection should use two-way verif
...
+**Server verifies client and client verifies server using Twisted and pyOpenSSL**
+
+.. code-block:: python
+
+    from OpenSSL import SSL, crypto
+    from cassandra.cluster import Cluster
+    from cassandra.io.twistedreactor import TwistedConnection
+
+    ssl_context = SSL.Context(SSL.TLSv1_METHOD)
+    ssl_context.set_verify(SSL.VERIFY_PEER, callback=lambda _1, _2, _3, _4, ok: ok)
```
Comment (aboudreault) [inline docs/security.rst]: just add the `ok` import

Comment (TheRealFalcon) [reply docs/security.rst]: There is no import. It's just the 5th argument to the lambda.

Comment (aboudreault) [reply docs/security.rst]: true :man_facepalming:
